samples/generic-synthetic-broken-links/brokenLinkChecker.js

```diff
@@ -0,0 +1,136 @@
+const puppeteer = require('puppeteer');
+
+/**
+ * Creates a dictionary of links that have a non-2xx status code (mapped to their respective status codes).
+ * @param {Array} An array of link objects with targetUrl and statusCode properties.
+ * @throws {Error} If any links have a non-2xx status code.
+ */
+function createNon2xxLinksDictionary(linksResult) {
+  const non2xxLinks = {};
+
+  for (const link of linksResult) {
+    const statusCode = link.statusCode;
+    const targetUrl = link.targetUrl;
+
+    if (statusCode < 200 || statusCode >= 300) {
+      non2xxLinks[targetUrl] = statusCode;
+    }
+  }
+
+  if (Object.keys(non2xxLinks).length === 0) {
+    return;
+  } else {
+    const stringified_dict = Object.entries(non2xxLinks).map(([key, value]) => `${key}:${value}`).join(", ");
+    throw new Error("Non 2xx Links: " + stringified_dict);
+  }
+}
+
+/**
+ * Retrieves all links on the page using Puppeteer, removing duplicate links and
+ * those ending with '#', to ensure reliable navigation within Puppeteer. Only follow links that start with http or https
+ *
+ * @param {Page} page - The Puppeteer page instance to retrieve the links from.
+ * @returns {Array<string>} - An array of the target URLs as strings.
+ */
+async function retrieveLinks(page) {
+  const links = await page.evaluate(() => {
+    const anchors = Array.from(document.querySelectorAll('a'));
```

Comment:
What sorcery is this, how is document referenced here when its not present anywhere else?